RHEL7 対応

xirc opened the pull request:
## 目的

RedHat Enterprise Linux 7 (RHEL7) 向けに 各種サービスをデプロイできるようになります。
EC2 の AMI として、CentOS8 に加えて、AWS Marketplace にある RHEL7.2 (`ami-0dd8f963`) が使えるようになります。
[AWS Marketplace: Red Hat Enterprise Linux (RHEL) 7.2 (HVM)](https://aws.amazon.com/marketplace/pp/B019NS7T5I?qid=1612505340203&sr=0-2&ref_=srh_res_product_title)

## 内容

- CentOS8 と RHEL7 の両方で動作するように次のモジュールのスクリプトを調整しました。
  - `modules/service/centos/core`
  - `modules/platform/aws/ec2`
- `example/aws_ec2`
  サンプルに RHEL7 を選べるように 変数 `bool use_rhel` を追加しました。  
  `use_rhel = true` を選ぶことで AMI として RHEL7 を 使えます。  
  デフォルトは `use_rhel = false` となっており、以前と同様に CentOS8 が使用されます。

## 設計上の懸念点

- ~~`modules/platform/centos` は *RedHat系のサポート* を表すような命名にしたほうがよいか？
  もしくは、rhel として別モジュールとする方がよいか？  
  現在の構造を保ったまま別モジュールに分離すると、
  CentOS と RHEL でほとんど同じようなコードとなることが問題になると思いました。~~  
  RHEL7 と CentOS でコードは共有します。  
  `modules/platform/centos` は、RedHatに対応されていることが分かりやすくなるように  `modules/platform/redhat` に変更します。

## 動作確認方法

`examples/aws_vpc` と `examples/aws_ec2` を使い、AWS に実際に環境を構築しました。
使用した AMI は `ami-0dd8f963` (RHEL7.2) です。
[Example Lerna Stack 構築ガイド](https://github.com/lerna-stack/lerna-terraform/tree/main/examples/aws_ec2#example-lerna-stack-%E6%A7%8B%E7%AF%89%E3%82%AC%E3%82%A4%E3%83%89)
に記載の手順で、環境構築と動作確認を実施しました。

Commit:
Revise AWS CLI install procedure in keepalived setup

- Use python36 instead of python38.
  python38 is not available for RHEL7.
- Don't change system the default python.
  yum installed in RHEL7, raises systex error because it is written in python2.
  Instead, run AWS CLI installer with python3 absolute path.

Commit:
Revise MariaDB's yum repository URL

The old URL is not available.

Commit:
Enable dnf tuning only if the os uses dnf as a package manager

RHEL7 uses yum instead of dnf as a package manager yet.

Commit:
Avoid an error if python is not registered in alternatives

RHEL7 does not register python in alternatives.
To works well in RHEL7 and CentOS8,
register python in alternatives in advance.

Commit:
Fix typo

SOCAT_PEERPORT would be better than SOCAT_PEER_PORT.
See https://manpages.debian.org/testing/socat/socat.1.en.html

Commit:
Avoid to happen a broken pipe error in fake-app

HAProxy may close a socket when the status code is returned by socat.
A broken pipe error happens since socat sends subsequent bytes to the
closed socket. To avoid the problem, socat will only send back the
status code.

Commit:
Disable appstream yum repository only if its repository file exists

Commit:
MariaDB yum repository file can be configured

A yum repository file that is used to install MariaDB, can be configured.
Repository files are provided by the module.
User can choose the file by specifying the distribution name among "centos8" or "rhel7".

Commit:
Change default keepalived_instance_role name

examples/aws_vpc generates an instance role whose name is keepalived_instance_role.
It would be better to match the name.

Commit:
aws_ec2 supports RHEL7

xirc commented:
@negokaz @tksugimoto レビューをお願いできますか？

negokaz commented on examples/aws_ec2/facility-lerna-stack-ec2.tf:
```diff
@@ -1,3 +1,27 @@
+locals {
+  dist_name = var.use_rhel ? "rhel7" : "centos8"
+  dist_conf = {
```
@xirc RHEL と CentOS でモジュールを共有することに関しては同意です。

サンプルに関しては理解しやすいようできるだけ条件分岐などを排除したシンプルな構造にしたほうが良いような気がしました。サンプルは CentOS を利用する前提の構成にしておき、RHEL を利用したいユーザーに向けてはその設定方法が書かれたガイドに誘導するという方法はどうでしょうか？

ガイドの中には AMI の変更方法と、その AMI の変更によって他のどの設定項目が影響を受けるか（ssh_user など）が解説されているイメージです。

xirc replied:
@negokaz 

レビューありがとうございます 😄 
サンプルをシンプルな構造に戻し、代わりにRHELの利用ガイドを追加しました。

> Revert "aws_ec2 supports RHEL7" … 5428104
> Add a document describing how to use RHEL7 9593549

CHANGELOG の更新を忘れていたため、このPRにて追加される内容を CHANGELOG に記載しました。
> Add RHEL7 support to the CHANGELOG 0c993f7

negokaz replied:
@xirc 

利用ガイド良いと思います！

さらに改善するなら、場合によってはインスタンスタイプを調整したいケースもあると思うので、「インスタンスタイプの設定」のセクションに「インスタンスタイプはマーケットプレイスに列挙されたもののみが利用できる」ということを示しておいたほうが、`c5` などを設定して apply がエラーになってしまうというのを回避できるので親切かなと思いました。私自身もここで躓いた経験があるので。

xirc replied:
@negokaz 

私も同様の経験をしたことを思い出しました :)
インスタンプタイプについての注意事項を記載します

xirc replied:
@negokaz 
利用ガイドを更新しました。
> Make the guide of RHEL7 more friendly … e66a114

negokaz commented:
@xirc 

> modules/platform/centos は RedHat系のサポート を表すような命名にしたほうがよいか？
> もしくは、rhel として別モジュールとする方がよいか？
> 現在の構造を保ったまま別モジュールに分離すると、
> CentOS と RHEL でほとんど同じようなコードとなることが問題になると思いました。

CentOS と RHEL で差分があまりないので、同じモジュールを共有する構成で問題ないと思います。

`modules/platform/centos` の命名は CentOS でしか使えない印象を与えてしまうような気がするので、変更したいですね。
個人的には一般的に CentOS などの RHEL から派生したディストリビューションは RedHat 系と呼ばれることが多い印象なので、`modules/platform/redhat` といった命名だと不自然さはないかなと思いました。

また、RHEL と CentOS の共通項である RPM を用いて `modules/platform/rpm` という命名も良さそうな気がしました。

Commit:
Revert "aws_ec2 supports RHEL7"

This reverts commit c9f8fcc57556ce4895f4fa305610ba2842d5b070.

Commit:
Add a document describing how to use RHEL7

Commit:
Add RHEL7 support to the CHANGELOG

xirc commented:
@negokaz 
> https://github.com/lerna-stack/lerna-terraform/pull/3#issuecomment-775595995
>
> modules/platform/centos の命名は CentOS でしか使えない印象を与えてしまうような気がするので、変更したいですね。
> 個人的には一般的に CentOS などの RHEL から派生したディストリビューションは RedHat 系と呼ばれることが多い印象なので、modules/platform/redhat といった命名だと不自然さはないかなと思いました。
>
> また、RHEL と CentOS の共通項である RPM を用いて modules/platform/rpm という命名も良さそうな気がしました。

私も変更するほうが良さそうに思います。
`modules/platform/rpm` も選択肢として良さそうですが、
個人的には、`modules/platform/redhat` の方が
「RedHat系に対応していること」ことが分かりやすいと思うため、
`modules/platform/redhat` を選択したいと思っています。
どう思いますか？

もしもモジュール名を変更する場合には、
「既に `modules/platform/centos` を使っているユーザがどのように `modules/platform/redhat` に移行するのか？」
について、検討が必要そうに思いました。

「terraformのモジュール名変更」
について私たちが気を付けた方が良いことはありますか？
私はモジュール名を変更することを経験したことがないため、
お持ちの知見を教えて頂けると嬉しいです。

`terraform state mv` を使ってモジュール名の変更ができるのではないか？と期待しています。
> terraform state mv
> https://www.terraform.io/docs/cli/commands/state/mv.html

negokaz commented:
@xirc 

https://github.com/lerna-stack/lerna-terraform/pull/3#issuecomment-785615012

> modules/platform/redhat を選択したいと思っています。
> どう思いますか？

`redhat` で良いと思います！

> もしもモジュール名を変更する場合には、
> 「既に modules/platform/centos を使っているユーザがどのように modules/platform/redhat に移行するのか？」
> について、検討が必要そうに思いました。
> 
> 「terraformのモジュール名変更」
> について私たちが気を付けた方が良いことはありますか？
> 私はモジュール名を変更することを経験したことがないため、
> お持ちの知見を教えて頂けると嬉しいです。

確かに移行方法についての検討は必要ですね。

モジュールのパスが変わったとしても、ユーザー側で定義した `module` の名前が変化しない限りはモジュール名変更にはならない認識です。

モジュールのパスが tfstate に入っていないかを気にしないといけないですが、類似の Terraform プロジェクトで apply 済みの tfstate ファイルを確認したところ `archive_file` data の中にモジュールのパスに依存するデータが含まれていました（zip ファイルに格納するディレクトリへのパスと zip ファイルの出力先のパス）。ただ、`archive_file` は設定ファイル全体のハッシュを取って変更があるかチェックする目的で定義していて、他のコンポーネントがこの zip ファイルのパスを参照することはないため、モジュールのパスが変わって zip ファイルの出力先が変わったとしても影響はないかなと思います。

また、`archive_file` 自体の ID は zip ファイルのハッシュから決まるみたいなので、設定ファイルの内容が変わってさえいなければこの `archive_file` の ID を trigger に設定しているリソースがあったとしても、余計な再実行が走るということもなさそうです。Go は読み書きの経験が 0 に近いので、予想を外してる可能性は高いですが…

> ```go
> 	d.SetId(d.Get("output_sha").(string))
> ```
> **[terraform-provider-archive/data_source_archive_file.go at main · hashicorp/terraform-provider-archive](https://github.com/hashicorp/terraform-provider-archive/blob/982982a1ae403caa653c53070b69ef6c379f3741/internal/provider/data_source_archive_file.go#L152)**

xirc commented:
@negokaz 

ありがとうございます！
私の理解では、ユーザは
"参照しているモジュール名を `modules/platform/centos` から `modules/platform/redhat` に 書き換える"
だけで大丈夫に思えてきましたが、移行作業を実施して確認してみます。

> モジュールのパスが tfstate に入っていないかを気にしないといけないですが、類似の Terraform プロジェクトで apply 済みの tfstate ファイルを確認したところ archive_file data の中にモジュールのパスに依存するデータが含まれていました（zip ファイルに格納するディレクトリへのパスと zip ファイルの出力先のパス）。ただ、archive_file は設定ファイル全体のハッシュを取って変更があるかチェックする目的で定義していて、他のコンポーネントがこの zip ファイルのパスを参照することはないため、モジュールのパスが変わって zip ファイルの出力先が変わったとしても影響はないかなと思います。

意図通りの差分がでることがありそうですね。
移行ガイドには、意図通りの差分がでることも記載したほうが良さそうです。

Commit:
Make the guide of RHEL7 more friendly

- Add a description about pricing information
- Add a description about available instance types

Commit:
Rename the modules from `service/centos` to `service/redhat`

- The modules now support Red Hat Enterprise Linux 7.
  - `service/centos/core` => `service/redhat/core`
  - `service/centos/dev` => `service/redhat/dev`
- The example `aws_ec2` uses new module names.
- Update documents to use new module names.

Commit:
The RHEL7 guide refers to each module by using its links

Commit:
Add a migration guide

xirc commented:
@negokaz 
モジュール名を `modules/service/centos` から `modules/service/redhat` に変更しました。  
合わせてマイグレーションガイドを作成しました。
ご確認よろしくお願いします 😄

negokaz commented on MIGRATION.md:
```diff
@@ -0,0 +1,208 @@
+
+#### `terraform.tfstate` にある旧モジュールへの参照を書き換える
+
+`terraform.tfstate` ファイルにあるファイルパスを書き換える必要があります。  
```
@xirc ファイルパスを書き換えないと、エラーなどが発生するということでしょうか？個人的にはモジュールの新しいパスが `tfstate` に反映されるだけで特にエラーなどは発生しないと予想していました。

xirc replied:
@negokaz 

`terraform.tfstate` にあるファイルパスを書き換えなくてもエラーは出ませんが、*モジュールのパス* が含まれるリソース(`management_script_global`と`cassandra_conf`)が`terraform plan`  での変更候補に出てきませんでした。旧パスが `terraform.tfstate` に残ってしまうとどのような問題があるのか不明であったため、安全のために、移行手順に `terraform.tfstate` の書き換えを記載しました。

`terraform.tfstate` を書き換えずに `terraform apply` を実施したことがなかったため、
頂いたコメントを確認後に、念のため`terraform.tfstate` を書き換えずに `terraform apply`を実施してみました。
`terraform plan` には候補として出てきませんが、もしかすると該当パスが新しいものに置き換わるのでは？と考えたためです。

実施してみると、該当パスが `terraform` によって正しく書き換えられていることが確認できました。

```
[***]$ grep -i 'modules/service/redhat' terraform.tfstate
            "output_path": "../../modules/service/redhat/core/resources/cassandra/conf.zip",
            "source_dir": "../../modules/service/redhat/core/resources/cassandra/conf",
            "output_path": "../../modules/service/redhat/core/resources/management-script/global.zip",
            "source_dir": "../../modules/service/redhat/core/resources/management-script/global",
```

結論として`terraform.tfstate` を書き換える作業は不要でした 😄 
`terraform.tfstate` を直接書き換える作業は可能な限り避けたいものであり、
不要であることも分かったため、移行手順からは削除します。

xirc replied:
@negokaz 移行手順を更新しました 😸 
> Remove the step of changing terraform state from MIGRATION guide … dccc504

negokaz replied:
@xirc 

LGTM!

Commit:
Remove the step of changing terraform state from MIGRATION guide

The step is not necessary for the migration.
`terraform apply` changes the path of modules even if it is not reported by `terraform plan`.

negokaz commented on README.md:
```diff
@@ -13,13 +13,13 @@ Service layer modules を単独で利用することで、プライベートク
+- **[service/redhat/core]**
     - 既存のサーバー上に Lerna Stack の各種サービス（Keepalived, Cassandra など）とアプリケーションをデプロイします
-- **[service/centos/dev]**
+- **[service/redhat/dev]**
```
@xirc 
モジュール名が変わったことにより、README を読んでも CentOS で使えるということが読み取れなくなってしまいましたね。

`service/redhat/core` と `service/redhat/dev` の説明部分に RHEL と CentOS の両方をサポートしているということを書いておくというのはどうでしょうか？

negokaz replied:
`platform/aws/ec2` の説明にも、デフォルトの OS が CentOS であることを書いておくと環境構築後の正しい構成をイメージしやすいかもしれませんね。

下の方に「RedHatEnterpriseLinux利用ガイド」の説明があるので、`platform/aws/ec2` はデフォルトで RHEL を構築するモジュールなのかな？と誤解される可能性もありそうだなと思いました。

xirc replied:
確かに次の2点が伝わりづらいですね。
- RHEL7, CentOS8 の両方をサポートしていること
- デフォルトが CentOS8 であること

明記するようにREADMEの記載を変更するほうが良さそうです 😄 
また、「RedHatEnterpriseLinux利用ガイド」をどのようなときに読むのかを README から説明すると、さらに良さそうに思いました。
変更します😄

xirc replied:
@negokaz 
次のコミットで変更しました。
> Revise README … 5d2e13a

negokaz replied:
LGTM! 🎉

Commit:
Revise README

Clarify
- these modules support CentOS8 and RHEL7.
- these modules use CentOS8 as a default.
- when to read the RHEL usage guide.
